src/Surfnet/Stepup/Identity/Identity.php

```diff
@@ -235,24 +238,74 @@ public function verifyEmail($verificationNonce)
         $secondFactorToVerify->verifyEmail();
     }
 
-    public function vetSecondFactor($registrationCode, $secondFactorIdentifier, $documentNumber, $identityVerified)
-    {
-        foreach ($this->verifiedSecondFactors as $secondFactor) {
-            if (!$secondFactor->wouldBeVettedBy(
-                $registrationCode,
-                $secondFactorIdentifier,
-                $documentNumber,
-                $identityVerified
-            )) {
-                continue;
-            }
+    public function vetSecondFactor(
+        IdentityApi $identity,
+        SecondFactorId $secondFactorId,
+        $secondFactorIdentifier,
+        $registrationCode,
+        $documentNumber,
+        $identityVerified
+    ) {
+        /** @var VettedSecondFactor|false $secondFactorWithHighestLoa */
+        $secondFactorWithHighestLoa = array_reduce(
+            $this->vettedSecondFactors->toArray(),
+            function (VettedSecondFactor $carry, VettedSecondFactor $item) {
+                return $carry->hasEqualOrHigherLoaComparedTo($item) ? $carry : $item;
+            },
+            $this->vettedSecondFactors->first()
+        );
 
-            $secondFactor->vet($registrationCode, $secondFactorIdentifier, $documentNumber, $identityVerified);
-            return;
+        $verifiedSecondFactorHasEqualOrLowerLoaComparedTo =
+            $identity->verifiedSecondFactorHasEqualOrLowerLoaComparedTo(
+                $secondFactorId,
+                $secondFactorWithHighestLoa
+            );
+
+        if (!$verifiedSecondFactorHasEqualOrLowerLoaComparedTo) {
+            throw new DomainException('Authority does not have the required LoA to vet the identity\'s second factor');
+        }
+
+        $identity->complyWithVettingOfSecondFactor(
+            $secondFactorId,
+            $secondFactorIdentifier,
+            $registrationCode,
+            $documentNumber,
+            $identityVerified
+        );
+    }
+
+    public function complyWithVettingOfSecondFactor(
+        SecondFactorId $secondFactorId,
+        $secondFactorIdentifier,
+        $registrationCode,
+        $documentNumber,
+        $identityVerified
+    ) {
+        if (!$this->verifiedSecondFactors->containsKey((string) $secondFactorId)) {
+            throw new DomainException('Cannot vet second factor: second factor does not exist.');
+        }
+
+        /** @var VerifiedSecondFactor $secondFactor */
+        $secondFactor = $this->verifiedSecondFactors->get((string) $secondFactorId);
+
+        if (!$secondFactor->wouldBeVettedBy(
+            $registrationCode,
+            $secondFactorIdentifier,
+            $documentNumber,
+            $identityVerified
+        )) {
+            throw new DomainException(
+                'Cannot vet second factor: registration code is incorrect or registration window has closed.'
+            );
```

Comment:
Either the checks should be done here and the DomainException should be made precise, of this should be guards of invariants in the SecondFactor and the specific exceptions should be thrown there. Reading the previous sentence to me there is only one solution, kinda like with the email verification. Thoughts?